### PR TITLE
Polymer 3 Conversion

### PR DESCRIPTION
**Pull request by dbatiste**
Not ready to merge yet - trying to iron out a couple of problems still.

### CODE REVIEW

File: d2l-input-checkbox-spacer.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+$_documentContainer.innerHTML = `<dom-module id="d2l-input-checkbox-spacer">
+	<template strip-whitespace="">
+		<style>
+			:host {
+				display: block;
+				padding-left: 1.7rem;
+				margin-bottom: 0.9rem;
+			}
+			:host-context([dir="rtl"]) {
```
Comment (dbatiste) [inline d2l-input-checkbox-spacer.js]: `:host(:dir(rtl))` approach needed here.

File: d2l-input-checkbox.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  then delete this comment!
+*/
+import '@polymer/polymer/polymer-legacy.js';
+
+import 'd2l-colors/d2l-colors.js';
+import 'd2l-typography/d2l-typography-shared-styles.js';
+import { Polymer } from '@polymer/polymer/lib/legacy/polymer-fn.js';
+// WORKAROUND: polymer-modulizer grabs non-existing Element export from polymer-element
+// TODO: Remove Element reference
+import { PolymerElement as Element } from '@polymer/polymer/polymer-element.js';
```
Comment (dbatiste) [inline d2l-input-checkbox.js]: Need to remove this and update the code that was referencing `Poilymer.Element`.

File: d2l-input-checkbox.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+			.d2l-input-checkbox-label {
+				@apply --d2l-body-compact-text;
+				color: var(--d2l-color-ferrite);
+				display: inline-block;
+				margin-left: 0.5rem;
+				vertical-align: middle;
+				white-space: normal;
+			}
+			:host-context([dir="rtl"]) .d2l-input-checkbox-label {
```
Comment (dbatiste) [inline d2l-input-checkbox.js]: This RTL code need to be updated in prep.

File: d2l-input-checkbox.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	},
+	focus: function() {
+		this.$$('input').focus();
+	},
+	_handleChange: function(e) {
+		this.checked = e.target.checked;
+
+		// in shady DOM the input's "change" event will leak through,
+		// so no need to fire it
+		if (Element || useShadow) {
```
Comment (dbatiste) [inline d2l-input-checkbox.js]: This needs to be updated since this logic no longer makes sense.

File: d2l-input-search.js [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+			input[type="text"].d2l-input {
+				font-family: inherit;
+				padding-right: 2.4rem;
+				overflow: hidden;
+				text-overflow: ellipsis;
+				white-space: nowrap;
+			}
+
+			:host-context([dir="rtl"]) input[type="text"].d2l-input {
```
Comment (dbatiste) [inline d2l-input-search.js]: This RTL code needs to be removed in patch.

File: d2l-input-search.js [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+				--d2l-button-icon-min-width: 1.7rem;
+				--d2l-button-icon-border-radius: 4px;
+				position: absolute;
+				height: 1.7rem;
+				width: 1.7rem;
+				top: 0.2rem;
+				right: 0.2rem;
+			}
+
+			:host-context([dir="rtl"]) d2l-button-icon {
```
Comment (dbatiste) [inline d2l-input-search.js]: These styles need to be removed so they don't bleed into LTR in Edge.

File: d2l-input-text.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+import '@polymer/polymer/polymer-legacy.js';
+
+import 'd2l-colors/d2l-colors.js';
+import 'd2l-polymer-behaviors/d2l-focusable-behavior.js';
+import './d2l-input-text-behavior.js';
+import './d2l-input-shared-styles.js';
+import { Polymer } from '@polymer/polymer/lib/legacy/polymer-fn.js';
+// WORKAROUND: polymer-modulizer grabs non-existing Element export from polymer-element
+// TODO: Remove Element reference
+import { PolymerElement as Element } from '@polymer/polymer/polymer-element.js';
```
Comment (dbatiste) [inline d2l-input-text.js]: This needs to be removed.

File: d2l-input-text.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+		/**
+		 * Value of the input.
+		 */
+		value: {
+			type: String,
+			notify: true
+		}
+	},
+	_handleChange: function() {
+		if (Element || useShadow) {
```
Comment (dbatiste) [inline d2l-input-text.js]: This logic needs to be updated since it no longer makes sense.

File: d2l-input-textarea.js [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+import 'fastdom/fastdom.js';
+import 'd2l-colors/d2l-colors.js';
+import 'd2l-polymer-behaviors/d2l-focusable-behavior.js';
+import './d2l-input-text-behavior.js';
+import './d2l-input-shared-styles.js';
+import { Polymer } from '@polymer/polymer/lib/legacy/polymer-fn.js';
+import { dom } from '@polymer/polymer/lib/legacy/polymer.dom.js';
+// WORKAROUND: polymer-modulizer grabs non-existing Element export from polymer-element
+// TODO: Remove Element reference
+import { PolymerElement as Element } from '@polymer/polymer/polymer-element.js';
```
Comment (dbatiste) [inline d2l-input-textarea.js]: This needs to be removed.

File: d2l-input-textarea.js [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+	},
+
+	_updateCached: function() {
+		fastdom.mutate(function() {
+			this.$.mirror.innerHTML = this._constrain(this.tokens);
+		}.bind(this), this);
+	},
+
+	_handleChange: function() {
+		if (Element || useShadow) {
```
Comment (dbatiste) [inline d2l-input-textarea.js]: This logic needs to be updated since it no longer makes sense.

File: demo/d2l-input-search.html [hunk 11]
```diff
@@ -4,66 +4,90 @@
...
-			});
-		</script>
-	</body>
-</html>
+			</demo-snippet>
+
+		</div>`;
+
+document.body.appendChild($_documentContainer.content);
+</script>cument.body.addEventListener('d2l-input-search-searched', function(e) {
```
Comment (dbatiste) [inline demo/d2l-input-search.html]: Something really strange has happened here.  Have not been able to figure this out yet.

File: demo/d2l-input-textarea.html [hunk 12]
```diff
@@ -137,17 +170,24 @@ <h3>d2l-input-textarea with change handler</h3>
...
+
+document.body.appendChild($_documentContainer.content);
+</script>
+		<script type="module">
+import '@polymer/iron-demo-helpers/demo-pages-shared-styles.js';
+import '@polymer/iron-demo-helpers/demo-snippet.js';
+import 'd2l-typography/d2l-typography.js';
+import '../d2l-input-textarea.js';
+// eslint-disable-next-line no-unused-vars
+function handleTextAreaChange(e) {
```
Comment (dbatiste) [inline demo/d2l-input-textarea.html]: This needs to be changed to wire-up via `addEventListener`.

File: demo/d2l-input-search.html (hunk 11 above)
Comment (margaree) [reply demo/d2l-input-search.html]: I've seen weirdness like this before if the line endings weren't LF

Comment (dbatiste) [reply demo/d2l-input-search.html]: @dlockhart @awikkerink : i've poked around in the code and i don't see the source of this.  Is this sort of thing something you've come across before?  The input (html) looks pretty clean from what i can tell.

Comment (dbatiste) [reply demo/d2l-input-search.html]: Thanks @margaree .  I will look into that.

Comment (dbatiste) [reply demo/d2l-input-search.html]: You were right about the line-endings!

**Pull request closed without merging**